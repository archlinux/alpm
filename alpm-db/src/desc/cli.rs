//! Command line interface setup for the `alpm-db-desc` executable.

use std::path::PathBuf;

use alpm_types::{
    Architecture,
    BuildDate,
    ExtraData,
    Group,
    InstalledSize,
    License,
    Name,
    OptionalDependency,
    PackageBaseName,
    PackageDescription,
    PackageInstallReason,
    PackageRelation,
    PackageValidation,
    Packager,
    Url,
    Version,
};
use clap::{Args, Parser, Subcommand, ValueEnum};
use strum::Display;

use crate::desc::schema::DbDescSchema;

/// The command-line interface handling for `alpm-db-desc`.
#[derive(Clone, Debug, Parser)]
#[command(
    about = "Command line tool to handle alpm-db-desc files",
    author,
    name = "alpm-db-desc",
    version
)]
pub struct Cli {
    /// The `alpm-db-desc` commands.
    #[command(subcommand)]
    pub command: Command,
}

/// The `alpm-db-desc` commands.
#[allow(clippy::large_enum_variant)]
#[derive(Clone, Debug, Subcommand)]
pub enum Command {
    /// Create a DB desc file according to a schema.
    #[command()]
    Create {
        /// The create subcommand.
        #[command(subcommand)]
        command: CreateCommand,
    },

    /// Validate a DB desc file according to a schema.
    #[command()]
    Validate {
        /// The validate arguments.
        #[command(flatten)]
        args: ValidateArgs,
    },

    /// Parse and output a DB desc file in a different format.
    #[command()]
    Format {
        /// The validate arguments.
        #[command(flatten)]
        args: ValidateArgs,

        /// Provide the output format.
        #[arg(
            short,
            long,
            value_name = "OUTPUT_FORMAT",
            default_value_t = OutputFormat::Json
        )]
        output_format: OutputFormat,

        /// Pretty-print the output.
        #[arg(short, long)]
        pretty: bool,
    },
}

/// Arguments for validating and parsing a DB desc file.
#[derive(Args, Clone, Debug)]
pub struct ValidateArgs {
    /// Provide the schema version to use.
    ///
    /// If not provided, the schema version will be detected from the file.
    #[arg(short, long, value_name = "VERSION")]
    pub schema: Option<DbDescSchema>,

    /// Provide the file to read.
    #[arg(value_name = "FILE")]
    pub file: Option<PathBuf>,
}

/// Arguments for creating a DB desc file according to the v1 schema.
#[derive(Args, Clone, Debug)]
pub struct V1CreateArgs {
    /// The package name.
    #[arg(env = "ALPM_DB_DESC_NAME", long)]
    pub name: Name,

    /// The package version.
    #[arg(env = "ALPM_DB_DESC_VERSION", long)]
    pub version: Version,

    /// The package base.
    #[arg(env = "ALPM_DB_DESC_BASE", long)]
    pub base: PackageBaseName,

    /// The package description.
    #[arg(env = "ALPM_DB_DESC_DESC", long)]
    pub description: Option<PackageDescription>,

    /// The package URL.
    #[arg(env = "ALPM_DB_DESC_URL", long)]
    pub url: Option<Url>,

    /// The package architecture.
    #[arg(env = "ALPM_DB_DESC_ARCH", long)]
    pub arch: Architecture,

    /// The package build date.
    #[arg(env = "ALPM_DB_DESC_BUILDDATE", long)]
    pub builddate: BuildDate,

    /// The package install date.
    #[arg(env = "ALPM_DB_DESC_INSTALLDATE", long)]
    pub installdate: BuildDate,

    /// The packager.
    #[arg(env = "ALPM_DB_DESC_PACKAGER", long)]
    pub packager: Packager,

    /// The installed size.
    #[arg(env = "ALPM_DB_DESC_SIZE", long)]
    pub size: InstalledSize,

    /// The package groups.
    #[arg(env = "ALPM_DB_DESC_GROUPS", long, value_delimiter = ' ')]
    pub groups: Vec<Group>,

    /// The package install reason.
    #[arg(env = "ALPM_DB_DESC_REASON", long)]
    pub reason: Option<PackageInstallReason>,

    /// The package licenses.
    #[arg(env = "ALPM_DB_DESC_LICENSE", long, value_delimiter = ' ')]
    pub license: Vec<License>,

    /// The package validation methods.
    #[arg(env = "ALPM_DB_DESC_VALIDATION", long, value_delimiter = ' ')]
    pub validation: Vec<PackageValidation>,

    /// The replaces.
    #[arg(env = "ALPM_DB_DESC_REPLACES", long, value_delimiter = ' ')]
    pub replaces: Vec<Name>,

    /// The dependencies.
    #[arg(env = "ALPM_DB_DESC_DEPENDS", long, value_delimiter = ' ')]
    pub depends: Vec<PackageRelation>,

    /// The optional dependencies.
    #[arg(env = "ALPM_DB_DESC_OPTDEPENDS", long, value_delimiter = ' ')]
    pub optdepends: Vec<OptionalDependency>,

    /// The conflicts.
    #[arg(env = "ALPM_DB_DESC_CONFLICTS", long, value_delimiter = ' ')]
    pub conflicts: Vec<Name>,

    /// The provides.
    #[arg(env = "ALPM_DB_DESC_PROVIDES", long, value_delimiter = ' ')]
    pub provides: Vec<Name>,

    /// The output file.
    #[arg(env = "ALPM_DB_DESC_OUTPUT_FILE")]
    pub output: Option<PathBuf>,
}

/// Create a DB desc file according to a schema.
#[derive(Clone, Debug, Subcommand)]
pub enum CreateCommand {
    /// Create a DB desc version 1 file.
    V1 {
        /// The create arguments.
        #[command(flatten)]
        args: V1CreateArgs,
    },

    /// Create a DB desc version 2 file.
    V2 {
        /// The create arguments.
        #[command(flatten)]
        args: V1CreateArgs,

        /// Structured extra metadata
        #[arg(
            env = "ALPM_DB_DESC_XDATA",
            long,
            value_delimiter = ' ',
            required = true
        )]
        xdata: Vec<ExtraData>,
    },
}

/// Output format for the format command.
#[derive(Clone, Debug, Default, Display, ValueEnum)]
pub enum OutputFormat {
    /// The JSON output format.
    #[default]
    #[strum(to_string = "json")]
    Json,
}
