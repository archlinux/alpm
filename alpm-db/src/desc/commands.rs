//! Commands for creating, validating, parsing, and formatting [alpm-db-desc] files.
//!
//! [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html

use std::{
    fs::{File, create_dir_all},
    io::{IsTerminal, Write, stdin},
};

use alpm_common::MetadataFile;

use crate::{
    Error,
    desc::{
        DbDescFile,
        DbDescFileV1,
        DbDescFileV2,
        cli::{CreateCommand, OutputFormat, ValidateArgs},
    },
};

/// Creates an [alpm-db-desc] file according to the specified version command.
///
/// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
///
/// # Notes
///
/// If no output file is specified, the data is printed to stdout.
///
/// # Errors
///
/// Returns an error if:
///
/// - the output file cannot be created,
/// - or the data cannot be written to the output file.
pub fn create_file(command: CreateCommand) -> Result<(), Error> {
    let (data, output) = match command {
        CreateCommand::V1 { args } => {
            let v1 = DbDescFileV1::new(
                args.name,
                args.version,
                args.base,
                args.description,
                args.url,
                args.arch,
                args.builddate,
                args.installdate,
                args.packager,
                args.size,
                args.groups,
                args.reason,
                args.license,
                args.validation,
                args.replaces,
                args.depends,
                args.optdepends,
                args.conflicts,
                args.provides,
            );
            (v1.to_string(), args.output)
        }
        CreateCommand::V2 { args, xdata } => {
            let v2 = DbDescFileV2::new(
                args.name,
                args.version,
                args.base,
                args.description,
                args.url,
                args.arch,
                args.builddate,
                args.installdate,
                args.packager,
                args.size,
                args.groups,
                args.reason,
                args.license,
                args.validation,
                args.replaces,
                args.depends,
                args.optdepends,
                args.conflicts,
                args.provides,
                xdata,
            );
            (v2.to_string(), args.output)
        }
    };

    if let Some(output_path) = output {
        // Create parent directories if necessary
        if let Some(output_dir) = output_path.parent() {
            create_dir_all(output_dir).map_err(|source| Error::IoPathError {
                path: output_dir.to_path_buf(),
                context: "creating output directory",
                source,
            })?;
        }

        let mut out = File::create(&output_path).map_err(|source| Error::IoPathError {
            path: output_path.clone(),
            context: "creating output file",
            source,
        })?;

        out.write_all(data.as_bytes())
            .map_err(|source| Error::IoPathError {
                path: output_path.clone(),
                context: "writing to output file",
                source,
            })?;
    } else {
        print!("{data}");
    }

    Ok(())
}

/// Parses an [alpm-db-desc] file and returns it as a [`DbDescFile`].
///
/// Data can be read from file or stdin.
///
/// # Errors
///
/// Returns an error if
///
/// - valid data cannot be read from a file,
/// - valid data cannot be read from stdin,
/// - or stdin is not a valid file descriptor/TTY.
///
/// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
pub fn parse(args: ValidateArgs) -> Result<DbDescFile, Error> {
    if let Some(file) = &args.file {
        DbDescFile::from_file_with_schema(file, args.schema)
    } else if !stdin().is_terminal() {
        DbDescFile::from_stdin_with_schema(args.schema)
    } else {
        Err(Error::NoInputFile)
    }
}

/// Validate a DB desc file by parsing it.
pub fn validate(args: ValidateArgs) -> Result<(), Error> {
    let _ = parse(args)?;
    Ok(())
}

/// Formats and optionally pretty prints an [alpm-db-desc] file.
///
/// # Errors
///
/// Returns an error if
///
/// - the [alpm-db-desc] file cannot be parsed,
/// - or the data cannot be formatted in the chosen output format.
///
/// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
pub fn format(args: ValidateArgs, output_format: OutputFormat, pretty: bool) -> Result<(), Error> {
    let desc = parse(args)?;
    match output_format {
        OutputFormat::Json => {
            let json = if pretty {
                serde_json::to_string_pretty(&desc).map_err(|e| Error::Json {
                    context: "serializing to pretty JSON",
                    source: e,
                })?
            } else {
                serde_json::to_string(&desc).map_err(|e| Error::Json {
                    context: "serializing to JSON",
                    source: e,
                })?
            };
            println!("{json}");
        }
    }

    Ok(())
}
