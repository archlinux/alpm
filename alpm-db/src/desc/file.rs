use std::{
    fmt::Display,
    fs::File,
    path::{Path, PathBuf},
    str::FromStr,
};

use alpm_common::{FileFormatSchema, MetadataFile};

use crate::{
    Error,
    desc::{DbDescFileV1, DbDescFileV2, DbDescSchema},
};

/// A representation of the [alpm-db-desc] file format.
///
/// Tracks all supported schema versions (`v1` and `v2`) of the database description file.
/// Each variant corresponds to a distinct layout of the format.
///
/// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
#[derive(Clone, Debug, PartialEq, serde::Serialize)]
#[serde(untagged)]
pub enum DbDescFile {
    /// The [alpm-db-descv1] file format.
    ///
    /// [alpm-db-descv1]: https://alpm.archlinux.page/specifications/alpm-db-descv1.5.html
    V1(DbDescFileV1),
    /// The [alpm-db-descv2] file format.
    ///
    /// This revision of the file format, adds the `%XDATA%` section.
    ///
    /// [alpm-db-descv2]: https://alpm.archlinux.page/specifications/alpm-db-descv2.5.html
    V2(DbDescFileV2),
}

impl MetadataFile<DbDescSchema> for DbDescFile {
    type Err = Error;

    /// Creates a [`DbDescFile`] from a file on disk, optionally validated using a [`DbDescSchema`].
    ///
    /// Opens the file and defers to [`DbDescFile::from_reader_with_schema`].
    ///
    /// # Examples
    ///
    /// ```
    /// use std::{fs::File, io::Write};
    ///
    /// use alpm_common::{FileFormatSchema, MetadataFile};
    /// use alpm_db::desc::{DbDescFile, DbDescSchema};
    /// use alpm_types::{SchemaVersion, semver_version::Version};
    ///
    /// # fn main() -> testresult::TestResult {
    /// // Prepare a file with DB desc data (v1)
    /// let (file, desc_data) = {
    ///     let desc_data = r#"%NAME%
    /// foo
    ///
    /// %VERSION%
    /// 1.0.0-1
    ///
    /// %BASE%
    /// foo
    ///
    /// %ARCH%
    /// x86_64
    ///
    /// %BUILDDATE%
    /// 1733737242
    ///
    /// %INSTALLDATE%
    /// 1733737243
    ///
    /// %PACKAGER%
    /// Foobar McFooface <foobar@mcfooface.org>
    ///
    /// %SIZE%
    /// 123
    ///
    /// "#;
    ///     let file = tempfile::NamedTempFile::new()?;
    ///     let mut output = File::create(&file)?;
    ///     write!(output, "{}", desc_data)?;
    ///     (file, desc_data)
    /// };
    ///
    /// let db_desc = DbDescFile::from_file_with_schema(
    ///     file.path(),
    ///     Some(DbDescSchema::V1(SchemaVersion::new(Version::new(1, 0, 0)))),
    /// )?;
    /// assert_eq!(db_desc.to_string(), desc_data);
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// Returns an error if:
    ///
    /// - the file cannot be opened for reading,
    /// - the contents cannot be parsed into any known [`DbDescFile`] variant,
    /// - or the provided [`DbDescSchema`] does not match the contents of the file.
    fn from_file_with_schema(
        file: impl AsRef<Path>,
        schema: Option<DbDescSchema>,
    ) -> Result<Self, Error> {
        let file = file.as_ref();
        Self::from_reader_with_schema(
            File::open(file).map_err(|source| Error::IoPathError {
                path: PathBuf::from(file),
                context: "opening the file for reading",
                source,
            })?,
            schema,
        )
    }

    /// Creates a [`DbDescFile`] from any readable stream, optionally validated using a
    /// [`DbDescSchema`].
    ///
    /// Reads the `reader` to a string buffer and defers to [`DbDescFile::from_str_with_schema`].
    ///
    /// # Examples
    ///
    /// ```
    /// use std::{fs::File, io::Write};
    ///
    /// use alpm_common::MetadataFile;
    /// use alpm_db::desc::{DbDescFile, DbDescSchema};
    /// use alpm_types::{SchemaVersion, semver_version::Version};
    ///
    /// # fn main() -> testresult::TestResult {
    /// // Prepare a reader with DB desc data (v2)
    /// let (reader, desc_data) = {
    ///     let desc_data = r#"%NAME%
    /// foo
    ///
    /// %VERSION%
    /// 1.0.0-1
    ///
    /// %BASE%
    /// foo
    ///
    /// %ARCH%
    /// x86_64
    ///
    /// %BUILDDATE%
    /// 1733737242
    ///
    /// %INSTALLDATE%
    /// 1733737243
    ///
    /// %PACKAGER%
    /// Foobar McFooface <foobar@mcfooface.org>
    ///
    /// %SIZE%
    /// 123
    ///
    /// %XDATA%
    /// pkgtype=pkg
    ///
    /// "#;
    ///     let file = tempfile::NamedTempFile::new()?;
    ///     let mut output = File::create(&file)?;
    ///     write!(output, "{}", desc_data)?;
    ///     (File::open(&file.path())?, desc_data)
    /// };
    ///
    /// let db_desc = DbDescFile::from_reader_with_schema(
    ///     reader,
    ///     Some(DbDescSchema::V2(SchemaVersion::new(Version::new(2, 0, 0)))),
    /// )?;
    /// assert_eq!(db_desc.to_string(), desc_data);
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// Returns an error if:
    ///
    /// - the `reader` cannot be read to string,
    /// - the data cannot be parsed into a known [`DbDescFile`] variant,
    /// - or the provided [`DbDescSchema`] does not match the parsed content.
    fn from_reader_with_schema(
        mut reader: impl std::io::Read,
        schema: Option<DbDescSchema>,
    ) -> Result<Self, Error> {
        let mut buf = String::new();
        reader
            .read_to_string(&mut buf)
            .map_err(|source| Error::IoReadError {
                context: "reading DB desc data",
                source,
            })?;
        Self::from_str_with_schema(&buf, schema)
    }

    /// Creates a [`DbDescFile`] from a string slice, optionally validated using a [`DbDescSchema`].
    ///
    /// If `schema` is [`None`], automatically infers the schema version by inspecting the input
    /// (`v1` = no `%XDATA%` section, `v2` = has `%XDATA%`).
    ///
    /// # Examples
    ///
    /// ```
    /// use alpm_common::MetadataFile;
    /// use alpm_db::desc::{DbDescFile, DbDescSchema};
    /// use alpm_types::{SchemaVersion, semver_version::Version};
    ///
    /// # fn main() -> testresult::TestResult {
    /// let v1_data = r#"%NAME%
    /// foo
    ///
    /// %VERSION%
    /// 1.0.0-1
    ///
    /// %BASE%
    /// foo
    ///
    /// %ARCH%
    /// x86_64
    ///
    /// %BUILDDATE%
    /// 1733737242
    ///
    /// %INSTALLDATE%
    /// 1733737243
    ///
    /// %PACKAGER%
    /// Foobar McFooface <foobar@mcfooface.org>
    ///
    /// %SIZE%
    /// 123
    ///
    /// "#;
    ///
    /// let dbdesc_v1 = DbDescFile::from_str_with_schema(
    ///     v1_data,
    ///     Some(DbDescSchema::V1(SchemaVersion::new(Version::new(1, 0, 0)))),
    /// )?;
    /// assert_eq!(dbdesc_v1.to_string(), v1_data);
    ///
    /// let v2_data = r#"%NAME%
    /// foo
    ///
    /// %VERSION%
    /// 1.0.0-1
    ///
    /// %BASE%
    /// foo
    ///
    /// %ARCH%
    /// x86_64
    ///
    /// %BUILDDATE%
    /// 1733737242
    ///
    /// %INSTALLDATE%
    /// 1733737243
    ///
    /// %PACKAGER%
    /// Foobar McFooface <foobar@mcfooface.org>
    ///
    /// %SIZE%
    /// 123
    ///
    /// %XDATA%
    /// pkgtype=pkg
    ///
    /// "#;
    ///
    /// let dbdesc_v2 = DbDescFile::from_str_with_schema(
    ///     v2_data,
    ///     Some(DbDescSchema::V2(SchemaVersion::new(Version::new(2, 0, 0)))),
    /// )?;
    /// assert_eq!(dbdesc_v2.to_string(), v2_data);
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// Returns an error if:
    ///
    /// - the input cannot be parsed into a valid [`DbDescFile`],
    /// - or the derived or provided schema does not match the detected format.
    fn from_str_with_schema(s: &str, schema: Option<DbDescSchema>) -> Result<Self, Error> {
        let schema = match schema {
            Some(schema) => schema,
            None => DbDescSchema::derive_from_str(s)?,
        };

        match schema {
            DbDescSchema::V1(_) => Ok(DbDescFile::V1(DbDescFileV1::from_str(s)?)),
            DbDescSchema::V2(_) => Ok(DbDescFile::V2(DbDescFileV2::from_str(s)?)),
        }
    }
}

impl Display for DbDescFile {
    /// Returns the textual representation of the [`DbDescFile`] in its corresponding
    /// [alpm-db-desc] format.
    ///
    /// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::V1(file) => write!(f, "{file}"),
            Self::V2(file) => write!(f, "{file}"),
        }
    }
}

impl FromStr for DbDescFile {
    type Err = Error;

    /// Creates a [`DbDescFile`] from a string slice.
    ///
    /// Internally calls [`DbDescFile::from_str_with_schema`] with `schema` set to [`None`].
    ///
    /// # Examples
    ///
    /// ```
    /// use std::str::FromStr;
    ///
    /// use alpm_db::desc::DbDescFile;
    ///
    /// # fn main() -> testresult::TestResult {
    /// let desc_data = r#"%NAME%
    /// foo
    ///
    /// %VERSION%
    /// 1.0.0-1
    ///
    /// %BASE%
    /// foo
    ///
    /// %ARCH%
    /// x86_64
    ///
    /// %BUILDDATE%
    /// 1733737242
    ///
    /// %INSTALLDATE%
    /// 1733737243
    ///
    /// %PACKAGER%
    /// Foobar McFooface <foobar@mcfooface.org>
    ///
    /// %SIZE%
    /// 123
    ///
    /// "#;
    ///
    /// let db_desc = DbDescFile::from_str(desc_data)?;
    /// assert_eq!(db_desc.to_string(), desc_data);
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// Returns an error if the input cannot be parsed into any known [`DbDescFile`] variant.
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::from_str_with_schema(s, None)
    }
}
