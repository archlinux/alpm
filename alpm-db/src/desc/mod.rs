//! Handling of [alpm-db-desc] file format versions.
//!
//! [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html

mod v1;
pub use v1::DbDescFileV1;

mod v2;
pub use v2::DbDescFileV2;

mod parser;
pub use parser::{Section, SectionKeyword};

mod schema;
pub use schema::DbDescSchema;

mod file;
pub use file::DbDescFile;

#[cfg(feature = "cli")]
pub mod cli;

#[cfg(feature = "cli")]
pub mod commands;
