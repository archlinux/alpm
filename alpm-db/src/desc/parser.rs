//! Parser for [alpm-db-desc] files.
//!
//! [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html

use std::{fmt::Display, str::FromStr};

use alpm_parsers::iter_str_context;
use alpm_types::{
    Architecture,
    BuildDate,
    ExtraData,
    Group,
    InstalledSize,
    License,
    Name,
    OptionalDependency,
    PackageBaseName,
    PackageDescription,
    PackageInstallReason,
    PackageRelation,
    PackageValidation,
    Packager,
    Url,
    Version,
};
use strum::{Display, EnumString, VariantNames};
use winnow::{
    ModalResult,
    Parser,
    ascii::{line_ending, newline, space0, till_line_ending},
    combinator::{
        alt,
        cut_err,
        delimited,
        eof,
        opt,
        peek,
        preceded,
        repeat,
        repeat_till,
        terminated,
    },
    error::{StrContext, StrContextValue},
    token::take_while,
};

/// A known section name in an [alpm-db-desc] file.
///
/// Section names are e.g. `%NAME%` or `%VERSION%`.
///
/// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
#[derive(Clone, Debug, Display, EnumString, Eq, Hash, PartialEq, VariantNames)]
#[strum(serialize_all = "UPPERCASE")]
pub enum SectionKeyword {
    /// %NAME%
    Name,
    /// %VERSION%
    Version,
    /// %BASE%
    Base,
    /// %DESC%
    Desc,
    /// %URL%
    Url,
    /// %ARCH%
    Arch,
    /// %BUILDDATE%
    BuildDate,
    /// %INSTALLDATE%
    InstallDate,
    /// %PACKAGER%
    Packager,
    /// %SIZE%
    Size,
    /// %GROUPS%
    Groups,
    /// %REASON%
    Reason,
    /// %LICENSE%
    License,
    /// %VALIDATION%
    Validation,
    /// %REPLACES%
    Replaces,
    /// %DEPENDS%
    Depends,
    /// %OPTDEPENDS%
    OptDepends,
    /// %CONFLICTS%
    Conflicts,
    /// %PROVIDES%
    Provides,
    /// %XDATA%
    XData,
}

impl SectionKeyword {
    /// Recognizes a [`SectionKeyword`] in an input string slice.
    ///
    /// # Examples
    ///
    /// ```
    /// use alpm_db::desc::SectionKeyword;
    ///
    /// # fn main() -> winnow::ModalResult<()> {
    /// let mut input = "%NAME%\nfoo\n";
    /// let kw = SectionKeyword::parser(&mut input)?;
    /// assert_eq!(kw, SectionKeyword::Name);
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// Returns an error if the input does not start with a valid
    /// `%SECTION%` header followed by a newline.
    pub fn parser(input: &mut &str) -> ModalResult<Self> {
        let section = delimited("%", take_while(1.., |c| c != '%'), "%");
        terminated(
            preceded(space0, section.try_map(Self::from_str)),
            line_ending,
        )
        .parse_next(input)
    }
}

/// A single logical section from a database desc file.
#[derive(Clone, Debug)]
pub enum Section {
    /// %NAME%
    Name(Name),
    /// %VERSION%
    Version(Version),
    /// %BASE%
    Base(PackageBaseName),
    /// %DESC%
    Desc(PackageDescription),
    /// %URL%
    Url(Url),
    /// %ARCH%
    Arch(Architecture),
    /// %BUILDDATE%
    BuildDate(BuildDate),
    /// %INSTALLDATE%
    InstallDate(BuildDate),
    /// %PACKAGER%
    Packager(Packager),
    /// %SIZE%
    Size(InstalledSize),
    /// %GROUPS%
    Groups(Vec<Group>),
    /// %REASON%
    Reason(PackageInstallReason),
    /// %LICENSE%
    License(Vec<License>),
    /// %VALIDATION%
    Validation(Vec<PackageValidation>),
    /// %REPLACES%
    Replaces(Vec<Name>),
    /// %DEPENDS%
    Depends(Vec<PackageRelation>),
    /// %OPTDEPENDS%
    OptDepends(Vec<OptionalDependency>),
    /// %CONFLICTS%
    Conflicts(Vec<Name>),
    /// %PROVIDES%
    Provides(Vec<Name>),
    /// %XDATA%
    XData(Vec<ExtraData>),
}

/// One or multiple newlines.
///
/// This also handles the case where there might be multiple lines with spaces.
fn newlines(input: &mut &str) -> ModalResult<()> {
    repeat(0.., line_ending).parse_next(input)
}

/// Parses a single scalar value from the input.
///
/// Consumes text until the end of the current line.
///
/// # Errors
///
/// Returns an error if the next token cannot be parsed into `T`.
fn value<T>(input: &mut &str) -> ModalResult<T>
where
    T: FromStr + Display,
    T::Err: Display,
{
    // Parse until the end of the line and attempt conversion to `T`.
    let value = till_line_ending.parse_to().parse_next(input)?;

    // Consume the newline or handle end-of-file gracefully.
    alt((line_ending, eof)).parse_next(input)?;

    Ok(value)
}

/// Parses a list of values from the input.
///
/// Repeats `value()` until the next section header (`%...%`)
/// or the end of the file.
///
/// # Errors
///
/// Returns an error if a value cannot be parsed into `T` or if the
/// section layout does not match expectations.
fn values<T>(input: &mut &str) -> ModalResult<Vec<T>>
where
    T: FromStr + Display,
    T::Err: Display,
{
    let next_section = peek(preceded(newline, SectionKeyword::parser)).map(|_| ());

    // Consume blank lines
    let blank_line = terminated(space0, newline).map(|_| ());

    repeat_till(0.., value, alt((next_section, blank_line, eof.map(|_| ()))))
        .context(StrContext::Label("values"))
        .context(StrContext::Expected(StrContextValue::Description(
            "expected a list of values in the database desc file",
        )))
        .parse_next(input)
        .map(|(outs, _)| outs)
}

/// Parses a single `%SECTION%` block and returns a [`Section`] variant.
///
/// # Errors
///
/// Returns an error if:
///
/// - the section name is invalid or not recognized,
/// - the section body contains malformed values,
/// - or the section does not terminate properly.
fn section(input: &mut &str) -> ModalResult<Section> {
    // Parse and validate the header keyword first.
    let section_keyword = cut_err(SectionKeyword::parser)
        .context(StrContext::Label("section name"))
        .context(StrContext::Expected(StrContextValue::Description(
            "expected a section name that is enclosed in `%` characters",
        )))
        .context_with(iter_str_context!([SectionKeyword::VARIANTS]))
        .parse_next(input)?;

    // Delegate to the corresponding value or values parser.
    let section = match section_keyword {
        SectionKeyword::Name => Section::Name(value(input)?),
        SectionKeyword::Version => Section::Version(value(input)?),
        SectionKeyword::Base => Section::Base(value(input)?),
        SectionKeyword::Desc => Section::Desc(value(input)?),
        SectionKeyword::Url => Section::Url(value(input)?),
        SectionKeyword::Arch => Section::Arch(value(input)?),
        SectionKeyword::BuildDate => Section::BuildDate(value(input)?),
        SectionKeyword::InstallDate => Section::InstallDate(value(input)?),
        SectionKeyword::Packager => Section::Packager(value(input)?),
        SectionKeyword::Size => Section::Size(value(input)?),
        SectionKeyword::Groups => Section::Groups(values(input)?),
        SectionKeyword::Reason => Section::Reason(value(input)?),
        SectionKeyword::License => Section::License(values(input)?),
        SectionKeyword::Validation => Section::Validation(values(input)?),
        SectionKeyword::Replaces => Section::Replaces(values(input)?),
        SectionKeyword::Depends => Section::Depends(values(input)?),
        SectionKeyword::OptDepends => Section::OptDepends(values(input)?),
        SectionKeyword::Conflicts => Section::Conflicts(values(input)?),
        SectionKeyword::Provides => Section::Provides(values(input)?),
        SectionKeyword::XData => Section::XData(values(input)?),
    };

    Ok(section)
}

/// Parses all `%SECTION%` blocks from the given input into a list of [`Section`]s.
///
/// This is the top-level parser used by the higher-level file constructors.
///
/// # Errors
///
/// Returns an error if:
///
/// - any section header is missing or malformed,
/// - a section value list fails to parse,
/// - or the overall structure of the file is inconsistent.
pub(crate) fn sections(input: &mut &str) -> ModalResult<Vec<Section>> {
    cut_err(repeat_till(
        0..,
        preceded(opt(newline), section),
        terminated(opt(newlines), eof),
    ))
    .context(StrContext::Label("sections"))
    .context(StrContext::Expected(StrContextValue::Description(
        "expected a section in the database desc file",
    )))
    .parse_next(input)
    .map(|(sections, _)| sections)
}
