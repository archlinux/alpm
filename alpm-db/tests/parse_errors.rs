//! Integration tests to ensure the desc parser produces meaningful errors.

use std::{fs::read_to_string, path::PathBuf, str::FromStr};

use alpm_db::desc::{DbDescFileV1, DbDescFileV2};
use insta::assert_snapshot;
use rstest::rstest;
use testresult::TestResult;

/// Each `.desc` file in `tests/parse_errors/` is expected to fail parsing.
///
/// Snapshots are saved in `parse_error_snapshots/` with the error message
/// and the file's contents as description.
#[rstest]
fn ensure_parse_errors(#[files("tests/parse_errors/*")] case: PathBuf) -> TestResult {
    let input = read_to_string(&case)?;
    let name = case
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("unknown_case");

    // Try both V1 and V2 constructors, as syntax errors should fail on either.
    let res_v1 = DbDescFileV1::from_str(&input);
    let res_v2 = DbDescFileV2::from_str(&input);

    // ensure at least one parser fails
    if res_v1.is_ok() || res_v2.is_ok() {
        return Err(format!("parser unexpectedly succeeded on {name}").into());
    }

    let error = res_v1
        .err()
        .map(|e| e.to_string())
        .or_else(|| res_v2.err().map(|e| e.to_string()))
        .unwrap_or_else(|| "unknown parser failure".to_string());

    let input_clone = input.clone();
    insta::with_settings!({
        description => input_clone,
        snapshot_path => "parse_errors/snapshots",
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(name, error);
    });

    Ok(())
}
