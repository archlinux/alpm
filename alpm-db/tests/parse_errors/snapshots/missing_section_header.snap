---
source: alpm-db/tests/parse_errors.rs
description: "NAME\nfoo\n\n%VERSION%\n1.0.0-1\n"
expression: error
---
Parser failed with the following error:
parse error at line 1, column 1
  |
1 | NAME
  | ^
invalid section name
expected expected a section name that is enclosed in `%` characters, `NAME`, `VERSION`, `BASE`, `DESC`, `URL`, `ARCH`, `BUILDDATE`, `INSTALLDATE`, `PACKAGER`, `SIZE`, `GROUPS`, `REASON`, `LICENSE`, `VALIDATION`, `REPLACES`, `DEPENDS`, `OPTDEPENDS`, `CONFLICTS`, `PROVIDES`, `XDATA`, expected a section in the database desc file
